detection_rules/eswrap.py

```diff
@@ -236,3 +241,182 @@ def collect_events(ctx, agent_hostname, index, agent_type, rta_name, rule_id, vi
+        try:
+            func(fid)
+        except elasticsearch.NotFoundError:
+            return False
+        if verbose:
+            click.echo(f' - {fid} deleted')
+        return True
+
+    model_exists = False
+    if not force:
+        existing_models = ml_client.get_trained_models()
+        model_exists = model_id in [m['model_id'] for m in existing_models.get('trained_model_configs', [])]
+
+    if model_exists or force:
+        if model_exists:
+            click.secho('[-] Existing model detected - deleting files', fg='yellow')
+
+        deleted = [
+            safe_delete(ingest_client.delete_pipeline, 'dga_ngram_expansion_inference'),
+            safe_delete(ingest_client.delete_pipeline, 'dns_classification_pipeline'),
+            safe_delete(es_client.delete_script, 'ngram-extractor-packetbeat'),
+            safe_delete(es_client.delete_script, 'ngram-remover-packetbeat'),
+            safe_delete(ml_client.delete_trained_model, model_id)
+        ]
+
+        if not any(deleted):
+            click.echo('No files deleted')
+    else:
+        click.echo(f'Model: {model_id} not found')
+
+
+@es_beta.command('setup-ml-dga')
+@click.option('--model-tag', '-t',
+              help='Release tag for model files staged in detection-rules (required to download files)')
+@click.option('--repo', '-r', default='elastic/detection-rules',
+              help='Repo hosting the model file releases (owner/repo)')
+@click.option('--model-dir', '-d', type=click.Path(exists=True, file_okay=False),
+              help='Directory containing local model files')
+@click.option('--overwrite', is_flag=True, help='Overwrite all files if already in the stack')
+@click.pass_context
+def setup_ml_dga(ctx, model_tag, repo, model_dir, overwrite):
+    """Upload ML DGA model and dependencies and enrich DNS data."""
+    import io
+    import requests
+    import shutil
+    import zipfile
+
+    es_client: Elasticsearch = ctx.obj['es']
+    client_info = es_client.info()
+
+    # download files if necessary
+    if not model_dir:
+        if not model_tag:
+            client_error('model-tag or model-dir required to download model files')
+
+        click.echo(f'Downloading artifact: {model_tag}')
+
+        release_url = f'https://api.github.com/repos/{repo}/releases/tags/{model_tag}'
+        release = requests.get(release_url)
+        release.raise_for_status()
+
+        zipped_url = release.json()['assets'][0]['browser_download_url']
+        zipped = requests.get(zipped_url)
+        z = zipfile.ZipFile(io.BytesIO(zipped.content))
+
+        dga_dir = get_path('ML-models', 'DGA')
+        model_dir = os.path.join(dga_dir, model_tag)
+        os.makedirs(dga_dir, exist_ok=True)
+        shutil.rmtree(model_dir, ignore_errors=True)
+        z.extractall(dga_dir)
+        click.echo(f'files saved to {model_dir}')
+
+        # read files as needed
+        z.close()
+
+    @contextmanager
+    def open_model_file(pattern, name_only=False):
```

Comment:
idk about the `name_only` pattern. sounds like two functions jammed into one, with conditional  return types

Reply:
💯 agree - I dislike it as well. I will just always return `name, contents` for consistency and unpack as needed